lib/cipherstash/index/bloom_filter.rb

```diff
@@ -1,28 +1,115 @@
-require "openssl"
+require_relative "../client/error"
 
 module CipherStash
   class Index
+    # A bloom filter implementation designed to be used with the *FilterMatch index classes.
+    #
+    # @private
     class BloomFilter
+      K_MIN = 3
+      K_MAX = 16
+      K_DEFAULT = 3
+      M_DEFAULT = 256
+      VALID_M_VALUES = [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536]
+
+      # The "set" bits of the bloom filter
       attr_reader :bits
+
+      # The size of the bloom filter in bits. Same as "filterSize" in the schema mapping and public docs.
+      #
+      #Since we only keep track of the set bits, the filter size determines the maximum value of the positions stored in the bits attr.
+      #
+      # Valid values are powers of 2 from 32 to 65536.
+      #
+      # @return [Integer]
+      attr_reader :m
+
+      # The number of hash functions applied to each term. Same as "filterTermBits" in the schema mapping and public docs.
+      #
+      # Implemented as k slices of a single hash.
+      #
+      # Valid values are integers from 3 to 16.
+      #
+      # @return [Integer]
       attr_reader :k
 
-      def initialize(key)
-        @key = key
+      # Creates a new bloom filter with the given key and filter match index settings.
+      #
+      # @param key [String] the key to use for hashing terms. Should be provided as a hex-encoded string.
+      #
+      # @param opts [Hash] the index settings.
+      #   "filterSize" and "filterTermBits" are used to set the m and k attrs respectively.
+      #
+      # @raise [CipherStash::Client::Error::InvalidSchemaError] if an invalid "filterSize" or "filterTermBits" is given.
+      def initialize(key, opts = {})
+        @key = [key].pack("H*")
+
+        unless @key.length == 32
+          raise ::CipherStash::Client::Error::InternalError, "expected bloom filter key to have length=32, got length=#{@key.length}"
+        end
+
         @bits = Set.new()
-        @k = 3
-      end
 
-      def add(term)
-        hash = OpenSSL::HMAC.digest("SHA256", @key, term)
+        @m = opts["filterSize"] || M_DEFAULT
+
+        unless VALID_M_VALUES.include?(@m)
+          raise ::CipherStash::Client::Error::InvalidSchemaError, "filterSize must be a power of 2 between 32 and 65536"
```

Comment:
Love these errors, they give *almost* all the info you need.  One suggested tweak:

```suggestion
          raise ::CipherStash::Client::Error::InvalidSchemaError, "filterSize must be a power of 2 between 32 and 65536 (got #{@m.inspect})"
```

The `inspect` call is an important part of the error, to ensure that if somehow a string representation of a number sneaks in (eg someone writes `"filterSize": "256"` in their JSON) it's possible to tell that it was a string that got passed in, not a number.  hashtag guess-how-I-learnt-this-lesson...